qiskit_experiments/database_service/db_experiment_data.py

```diff
@@ -910,16 +957,57 @@ def load(cls, experiment_id: str, service: DatabaseServiceV1) -> "DbExperimentDa
         expdata._created_in_db = True
         return expdata
 
-    def cancel_jobs(self) -> None:
-        """Cancel any running jobs."""
-        for kwargs, fut in self._job_futures.copy():
-            if not fut.done():
-                for job in kwargs["jobs"]:
-                    if job.status() not in JOB_FINAL_STATES:
-                        try:
-                            job.cancel()
-                        except Exception as err:  # pylint: disable=broad-except
-                            LOG.info("Unable to cancel job %s: %s", job.job_id(), err)
+    def jobs(self) -> List[Job]:
+        """Return a list of jobs for the experiment"""
+        return self._jobs.values()
+
+    def cancel_jobs(self) -> bool:
+        """Cancel any running jobs.
+
+        Returns:
+            True if all jobs successfully cancelled, otherwise false.
+        """
+        cancelled = True
+        with self._jobs.lock:
+            for job in self._jobs.values():
+                if job and job.status() not in JOB_FINAL_STATES:
+                    try:
+                        job.cancel()
```

Comment:
What will happen to thread pool executor? Is it terminated by error without releasing resource?

Reply:
I had to make some changes for next commit, but if the job is cancelled, the `job.result()` call in the future function will raise an exception, which will be caught and handled so that the future finishes and logs that the job was cancelled.